compiler/src/dotty/tools/dotc/transform/Splicing.scala

```diff
@@ -0,0 +1,394 @@
+ *  Those references appear in quotes that are nested in a splice.
+ *
+ *  After this phase we have the invariant where all splices have the following shape
+ *  ```
+ *  {{{ <holeIdx> | <holeType> | <captures>* | (<capturedTerms>*) => <spliceContent> }}}
+ *  ```
+ *  where `<spliceContent>` does not contain any free references to quoted definitions and `<captures>*`
+ *  contains the quotes with references to all cross-quote references. There are some special rules
+ *  for references in the LHS of assignments and cross-quote method references.
+ *
+ *  In the following code example `x1` and `x2` are cross-quote references.
+ *  ```
+ *  '{ ...
+ *    val x1: T1 = ???
+ *    val x2: T2 = ???
+ *    ${ (q: Quotes) ?=> f('{ g(x1, x2) }) }: T3
+ *  }
+ *  ```
+ *
+ *  This phase identifies cross-quote references such as `x1` and replaces it with an `${x1$}`.
+ *  All cross-quote arguments are directly applied in the lambda.
+ *
+ *  ```
+ *  '{ ...
+ *     val x1: T1 = ???
+ *     val x2: T2 = ???
+ *     {{{ 0 | T3 | x1, x2 |
+ *       (x1$: Expr[T1], x2$: Expr[T2]) => // body of this lambda does not contain references to x1 or x2
+ *         (q: Quotes) ?=> f('{ g(${x1$}, ${x2$}) })
+ *
+ *     }}}
+ *   }
+ *   ```
+ *
+ *  and then performs the same transformation on `'{ g(${x1$}, ${x2$}) }`.
+ *
+ */
+class Splicing extends MacroTransform:
+  import tpd._
+
+  override def phaseName: String = Splicing.name
+
+  override def run(using Context): Unit =
+    if (ctx.compilationUnit.needsQuotePickling) super.run(using freshStagingContext)
+
+  protected def newTransformer(using Context): Transformer = Level0QuoteTransformer
+
+  /** Transforms all quotes at level 0 using the `QuoteTransformer` */
+  private object Level0QuoteTransformer extends Transformer:
+    override def transform(tree: tpd.Tree)(using Context): tpd.Tree =
+      assert(level == 0)
+      tree match
+        case Apply(Select(Apply(TypeApply(fn,_), List(code)),nme.apply),List(quotes))
+        if fn.symbol == defn.QuotedRuntime_exprQuote =>
+          QuoteTransformer().transform(tree)
+        case TypeApply(_, _) if tree.symbol == defn.QuotedTypeModule_of =>
+          QuoteTransformer().transform(tree)
+        case tree: DefDef if tree.symbol.is(Inline) =>
+          // Quotes in inlined methods are only pickled after they are inlined.
+          tree
+        case _ =>
+          super.transform(tree)
+  end Level0QuoteTransformer
+
+
+  /** Transforms all direct splices in the current quote and replace them with holes. */
+  private class QuoteTransformer() extends Transformer:
+    /** Set of definitions in the current quote */
+    private val quotedDefs = mutable.Set.empty[Symbol]
+
+    /** Number of holes created in this quote. Used for indexing holes. */
+    private var numHoles = 0
+
+    /** Mapping from the term symbol of a `Type[T]` to it's hole. Used to deduplicate type holes. */
+    private var typeHoles = mutable.Map.empty[Symbol, Hole]
```

Comment:
This also probably can be a val and not a var since it's mutable.

Reply:
Changed to `val`